In [4]:
import pandas as pd
import glob

# Define the directory where your CSV files are stored
csv_directory = "./PassengerVehicle_Stats/"  # Replace with the directory path

# Use glob to load all CSV files in the directory into a list
csv_files = glob.glob(csv_directory + "*.csv")
try:
        # Check if we have 9 files and load them into a list
    if len(csv_files) == 9:
        print(f"Found {len(csv_files)} CSV files.")
        
        # Load all CSV files into a list of DataFrames
        csv_data_list = [pd.read_csv(file) for file in csv_files]

        # Concatenate the list of DataFrames into a single DataFrame
        vehicles_df = pd.concat(csv_data_list, ignore_index=True)

        # Optionally, save the combined DataFrame to a new CSV file
        vehicles_df.to_csv("combined_vehicles_data.csv", index=False)
        print("Combined DataFrame saved to 'combined_vehicles_data.csv'.")
    else:
        print(f"Expected 9 CSV files, but found {len(csv_files)}. Please check your directory.")

except Exception as e:
    print(f"Error occurred: {e}")

ModuleNotFoundError: No module named 'pandas'

In [ ]:
#Renaming the first column
vehicles_df.rename(columns={"Unnamed: 0": "Index"}, inplace=True)
print(vehicles_df.columns)

Index(['Index', 'Public Vehicle Number', 'Status', 'Vehicle Make',
       'Vehicle Model', 'Vehicle Model Year', 'Vehicle Color',
       'Vehicle Fuel Source', 'Wheelchair Accessible', 'Company Name',
       'Address', 'City', 'State', 'ZIP Code', 'Taxi Affiliation',
       'Taxi Medallion License Management ', 'Record ID'],
      dtype='object')


In [ ]:
#Data types of the variables
print(vehicles_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16602 entries, 0 to 16601
Data columns (total 17 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Unnamed: 0                          16602 non-null  int64  
 1   Public Vehicle Number               16602 non-null  int64  
 2   Status                              16602 non-null  object 
 3   Vehicle Make                        14685 non-null  object 
 4   Vehicle Model                       14639 non-null  object 
 5   Vehicle Model Year                  14660 non-null  float64
 6   Vehicle Color                       14616 non-null  object 
 7   Vehicle Fuel Source                 16602 non-null  object 
 8   Wheelchair Accessible               16602 non-null  object 
 9   Company Name                        16602 non-null  object 
 10  Address                             14816 non-null  object 
 11  City                                14816

In [ ]:
#First few records of the vehicles data frame
print(vehicles_df.head())

   Unnamed: 0  Public Vehicle Number    Status Vehicle Make Vehicle Model  \
0        1286                  12009  RESERVED    CHEVROLET       EXPRESS   
1        2095                  12248  INACTIVE     MERCEDES      SPRINTER   
2        7950                  13527  INACTIVE     VAN HOOL         TD925   
3        8700                  12248  INACTIVE     MERCEDES      SPRINTER   
4        9359                  13528  INACTIVE     VAN HOOL         TD925   

   Vehicle Model Year Vehicle Color Vehicle Fuel Source Wheelchair Accessible  \
0              2014.0         BLACK          Bio-Diesel                     N   
1              2010.0        SILVER          Bio-Diesel                     N   
2              2008.0           RED          Bio-Diesel                     N   
3              2010.0        SILVER          Bio-Diesel                     N   
4              2008.0           RED          Bio-Diesel                     N   

                              Company Name        

In [ ]:
# All the variables consisit with qualitative nature
#Identifying the mode of variables
qualitative_columns = vehicles_df.select_dtypes(include=['object', 'category']).columns
mode_values = vehicles_df[qualitative_columns.drop('Record ID')].mode()
print("Mode values for each qualitative variable:")
print(mode_values)

Mode values for each qualitative variable:
   Status Vehicle Make Vehicle Model Vehicle Color Vehicle Fuel Source  \
0  ACTIVE       TOYOTA         CAMRY         WHITE            Gasoline   

  Wheelchair Accessible          Company Name                 Address  \
0                     N  CITY OF CHICAGO BACP  3800 N. MILWAUKEE AVE.   

      City State Taxi Affiliation Taxi Medallion License Management   
0  CHICAGO    IL   DIS-AFFILIATED                      Owner Manager  


In [ ]:
#obtaining counts of each status
status_category_counts = vehicles_df['Status'].value_counts()

print(status_category_counts)

Status
ACTIVE         7222
VIOLATION      3974
INACTIVE       3702
SURRENDER       888
RESERVED        582
REVOKED         147
FORECLOSURE      82
HOLD              5
Name: count, dtype: int64


In [ ]:
#Wheel chair accessibility 
wheelchair_category_counts = vehicles_df['Wheelchair Accessible'].value_counts()

print(wheelchair_category_counts)

Wheelchair Accessible
N    15978
Y      624
Name: count, dtype: int64


In [ ]:
#Fuel type variation
fuel_category_counts = vehicles_df['Vehicle Fuel Source'].value_counts()

print(fuel_category_counts)

Vehicle Fuel Source
Gasoline                  9938
Hybrid                    5969
Pedal                      215
Flex Fuel                  192
Diesel                      96
Compressed Natural Gas      91
Horse                       59
Electric                    33
Bio-Diesel                   9
Name: count, dtype: int64


In [ ]:
#Detecting Duplicates
duplicates = vehicles_df[vehicles_df.duplicated(subset='Record ID', keep=False)]
len(duplicates)


1838

In [ ]:
#Removing Duplicates
vehicles_df = vehicles_df.drop_duplicates(subset=['Record ID'], keep='first')
print(vehicles_df)
vehicles_df.to_csv('cleaned_vehicles.csv', index=False)

       Unnamed: 0  Public Vehicle Number     Status Vehicle Make  \
0            1286                  12009   RESERVED    CHEVROLET   
1            2095                  12248   INACTIVE     MERCEDES   
2            7950                  13527   INACTIVE     VAN HOOL   
4            9359                  13528   INACTIVE     VAN HOOL   
5            9441                  12025   INACTIVE     MERCEDES   
...           ...                    ...        ...          ...   
16594       16033                    166     ACTIVE      MANMADE   
16595       16048                    194   INACTIVE   MAINSTREET   
16597       16143                     42     ACTIVE    PRECISION   
16599       16216                     70  VIOLATION   MAINSTREET   
16601       16281                    117     ACTIVE      SCHWINN   

      Vehicle Model  Vehicle Model Year Vehicle Color Vehicle Fuel Source  \
0           EXPRESS              2014.0         BLACK          Bio-Diesel   
1          SPRINTER          

In [ ]:
#Identifying null value percentages of variables
missing_percentage = (vehicles_df.isnull().sum() / len(vehicles_df)) * 100
print(missing_percentage)

In [ ]:
#Import matplotlib
import matplotlib.pyplot as plt


In [ ]:
#Plot Null value percentages
plt.figure(figsize=(8, 5))
missing_percentage.sort_values(ascending=False).plot(kind='bar', color='skyblue', edgecolor='black')
plt.title('Percentage of Null Values by Column', fontsize=14)
plt.ylabel('Percentage (%)', fontsize=12)
plt.xlabel('Columns', fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()


Since there are more than 50% of null values for "Taxi Medallion License Management" and "Taxi Affiliation", those two variables should removed.

In [ ]:
#Removing "Taxi Medallion License Management" and "Taxi Affiliation"  columns from the dataset

columns_to_drop = ["Taxi Affiliation", "Taxi Medallion License Management"] 
vehicles_df = vehicles_df.drop(columns=columns_to_drop, errors='ignore')


In [ ]:
# Save the updated DataFrame to a new CSV file
vehicles_df.to_csv("updated_combined_vehicles_data.csv", index=False)
print("Updated DataFrame saved to 'updated_combined_vehicles_data.csv'.")

In [ ]:
#count the number of unique values for each column in the dataframe
vehicles_df.nunique()

Index                                 14473
Public Vehicle Number                 14473
Status                                    8
Vehicle Make                             78
Vehicle Model                           276
Vehicle Model Year                       47
Vehicle Color                            81
Vehicle Fuel Source                       8
Wheelchair Accessible                     2
Company Name                           7084
Address                                2239
City                                     37
State                                     1
ZIP Code                                 92
Taxi Affiliation                         24
Taxi Medallion License Management        10
Record ID                             14473
dtype: int64

In [ ]:
#count number of null values in the columns of the dataframe
vehicles_df.isnull().sum()

Index                                    0
Public Vehicle Number                    0
Status                                   0
Vehicle Make                          1801
Vehicle Model                         1837
Vehicle Model Year                    1819
Vehicle Color                         1855
Vehicle Fuel Source                      0
Wheelchair Accessible                    0
Company Name                             0
Address                               1688
City                                  1688
State                                 1688
ZIP Code                              1688
Taxi Affiliation                      8091
Taxi Medallion License Management     8114
Record ID                                0
dtype: int64

Here we are dropping rows where the following columns contain missing values (`NaN`):
- `Address`: Contains 1688 missing values.
- `City`: Contains 1688 missing values.
- `State`: Contains 1688 missing values.
- `ZIP Code`: Contains 1688 missing values.
- `Vehicle Model Year`: Contains 1819 missing values.

These columns are critical for analysis, so rows with incomplete data in these fields are removed.

In [ ]:
vehicles_df.dropna(subset= ['Address','City','State', 'ZIP Code', 'Vehicle Model Year'], inplace = True)

In [ ]:
# Save the cleaned DataFrame to a CSV file
vehicles_df.to_csv("rows_drop_1_vehicles.csv", index=False)

print("Cleaned DataFrame saved to 'rows_drop_1_vehicles.csv'.")
print(vehicles_df)

Cleaned DataFrame saved to 'rows_drop_1_vehicles.csv'.
       Index  Public Vehicle Number     Status Vehicle Make Vehicle Model  \
0       1286                  12009   RESERVED    CHEVROLET       EXPRESS   
1       2095                  12248   INACTIVE     MERCEDES      SPRINTER   
2       7950                  13527   INACTIVE     VAN HOOL         TD925   
4       9359                  13528   INACTIVE     VAN HOOL         TD925   
5       9441                  12025   INACTIVE     MERCEDES      SPRINTER   
...      ...                    ...        ...          ...           ...   
16382  16576                    998  VIOLATION       TOYOTA         CAMRY   
16383  16583                   3810     ACTIVE       TOYOTA          RAV4   
16384  16590                   4690  VIOLATION       TOYOTA         CAMRY   
16385  16592                   2831     ACTIVE       TOYOTA         PRIUS   
16386  16596                   1509     ACTIVE       TOYOTA        SIENNA   

       Vehicle Model